<a href="https://colab.research.google.com/github/purnawanpp/plants_book/blob/main/plants.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Protein-Ligand ANTSystem (PLANTS) simulation by Purnawan Pontana Putra**

**Required software installation**

In [5]:
!wget https://github.com/purnawanpp/plants/raw/main/plants
!wget https://github.com/purnawanpp/plants/raw/main/spores
!apt install openbabel
!chmod +x /content/plants
!chmod +x /content/spores
!wget https://github.com/purnawanpp/plants/raw/main/dock_gcolab.txt

--2023-09-29 23:03:19--  https://github.com/purnawanpp/plants/raw/main/plants
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/purnawanpp/plants/main/plants [following]
--2023-09-29 23:03:19--  https://raw.githubusercontent.com/purnawanpp/plants/main/plants
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9536847 (9.1M) [application/octet-stream]
Saving to: ‘plants’

plants              100%[===================>]   9.09M  --.-KB/s    in 0.08s   

2023-09-29 23:03:20 (120 MB/s) - ‘plants’ saved [9536847/9536847]

--2023-09-29 23:03:20--  https://github.com/purnawanpp/plants/raw/main/spores
Resolvin

**Download Protein PDB ID 3HTB**

In [6]:
!wget https://files.rcsb.org/download/3HTB.pdb

--2023-09-29 23:03:34--  https://files.rcsb.org/download/3HTB.pdb
Resolving files.rcsb.org (files.rcsb.org)... 128.6.159.100, 128.6.159.157, 132.249.213.241, ...
Connecting to files.rcsb.org (files.rcsb.org)|128.6.159.100|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/octet-stream]
Saving to: ‘3HTB.pdb’

3HTB.pdb                [  <=>               ] 161.21K   609KB/s    in 0.3s    

2023-09-29 23:03:35 (609 KB/s) - ‘3HTB.pdb’ saved [165078]



**Separete Protein and ligand**

This mode will perform a structure recognition and protonation, the output are written in files namely protein.mol2
which contains all protein chains from the input, ligand ligandname_X.mol2 which
contain the ligand molecuels (ligand name is generated from the residue name of
the ligand in the original file, and X is a running number in case there are two
ligands with identical names) and water.mol2 which contains all water molecules
from the input file.

In [7]:
!./spores --mode splitpdb 3HTB.pdb


				 SPORES        

	      Structure PrOtonation and REcognition System

			      version 1.3


author: T. ten Brink

scientific contributors: P. Duchstein, J. Gimmler, O. Korb, M. Zerara,
			  T.E. Exner

contact: tim.tenbrink@uni-konstanz.de


molecule number: 1
64 alternative atomic positions have been removed !
splitting 3HTB
225
protein: 0
ligs: 4
water: 660
metal: 0


**Kode Native Ligand protein 3HTB adalah JZ4, baca website https://www.rcsb.org/structure/3HTB untuk mengetahui native ligannya



In [8]:
!mv ligand_JZ4167_0.mol2 lig.mol2
!mv protein.mol2 rec.mol2

**Create and define bindingsite or Gridbox**

In [10]:
!./plants --mode bind lig.mol2 rec.mol2



                  PLANTS
        Protein-Ligand ANT System
               version 1.2


author: Oliver Korb

scientific contributors: T.E. Exner, T. Stuetzle

contact: Oliver.Korb@uni-konstanz.de


run PLANTS: PLANTS --mode screen yourconfigfile



Calculated bindingsite definition (from 1 ligands): 

bindingsite_center 22.7142 -25.2627 -3.283
bindingsite_radius 3.75624



Move data:

* bindingsite_center 22.7142 -25.2627 -3.283
* bindingsite_radius 3.75624

into the file **dock_gcolab.txt**


**Perintah untuk menginput binding site kedalam script**

In [14]:
# Baca nilai bindingsite_center dan bindingsite_radius dari file bindingsite.def
with open('/content/bindingsite.def', 'r') as bindingsite_file:
    for line in bindingsite_file:
        if line.startswith('bindingsite_center'):
            bindingsite_center_baru = line.strip().split(' ')[1:]
            bindingsite_center_baru = ' '.join(bindingsite_center_baru)
        elif line.startswith('bindingsite_radius'):
            bindingsite_radius_baru = line.strip().split(' ')[1]

# Baca isi file dock_gcolab.txt dan ganti nilai bindingsite_center dan bindingsite_radius
with open('/content/dock_gcolab.txt', 'r') as dock_file:
    lines = dock_file.readlines()

with open('/content/dock_gcolab.txt', 'w') as dock_file:
    for line in lines:
        if line.startswith('bindingsite_center'):
            dock_file.write(f'bindingsite_center {bindingsite_center_baru}\n')
        elif line.startswith('bindingsite_radius'):
            dock_file.write(f'bindingsite_radius {bindingsite_radius_baru}\n')
        else:
            dock_file.write(line)

# Baca isi file dock_comp.txt dan ganti nilai bindingsite_center dan bindingsite_radius
with open('/content/dock_comp.txt', 'r') as dock_file:
    lines = dock_file.readlines()

with open('/content/dock_comp.txt', 'w') as dock_file:
    for line in lines:
        if line.startswith('bindingsite_center'):
            dock_file.write(f'bindingsite_center {bindingsite_center_baru}\n')
        elif line.startswith('bindingsite_radius'):
            dock_file.write(f'bindingsite_radius {bindingsite_radius_baru}\n')
        else:
            dock_file.write(line)


**Perintah melakukan molecular docking**

In [ ]:
!./plants --mode screen /content/dock_gcolab.txt



                  PLANTS
        Protein-Ligand ANT System
               version 1.2


author: Oliver Korb

scientific contributors: T.E. Exner, T. Stuetzle

contact: Oliver.Korb@uni-konstanz.de


run PLANTS: PLANTS --mode screen yourconfigfile

PLANTS info: CHO hydrogen found 513 HD2 HIS31
PLANTS info: CHO hydrogen found 514 HE1 HIS31
virtual screening progress: 1 of 1
current ligand: /content/lig.mol2 (entry 1)
LIGAND DOFs: 9
PROTEIN DOFs: 0
Simplex dimension: 9
Simplex dimension: 9
iterations : 187
starting optimization ...
problem dimension: 9
ATOMS / s: 6.93417e+06
EVAL / s: 693417
optimization finished after 1.07s
best score: -67.44


total virtual screening time: 1.25s
Ligands skipped due to errors: 0


**The ΔG scores or docking results**

After the process is complete, don't forget to click the Google Colab refresh box/file icon. Docking simulation results can be seen in the *results* folder with the file name **bestranking.csv**, see energy TOTAL_SCORE is a delta G value

In [ ]:
import pandas as pd
df = pd.read_csv('/content/results/bestranking.csv')
df.head()

LIGAND_ENTRY  TOTAL_SCORE  SCORE_RB_PEN  \
0  3HTB_ligand_2_entry_00001_conf_01     -67.4443      -61.4443   

   SCORE_NORM_HEVATOMS  SCORE_NORM_CRT_HEVATOMS  SCORE_NORM_WEIGHT  \
0             -6.74443                 -31.3049          -0.495237   

   SCORE_NORM_CRT_WEIGHT  SCORE_RB_PEN_NORM_CRT_HEVATOMS  SCORE_NORM_CONTACT  \
0               -13.1089                        -28.5199            -6.74443   

     EVAL      TIME  
0  732254  0.846584

Explanation:

* LIGAND_ENTRY: Name or identification code of the ligand that was docked.
* TOTAL_SCORE: Overall score generated from docking the ligand to the target protein.
* SCORE_RB_PEN: Score for rotamer penalty.
* SCORE_NORM_HEVATOMS: Score for heavy heteroatoms (in this case, normalized value).
* SCORE_NORM_CRT_HEVATOMS: Score for normalized heavy heteroatoms with correction.
* SCORE_NORM_WEIGHT: Score normalized by the weight of the ligand molecule.
* SCORE_NORM_CRT_WEIGHT: Score normalized by correction and weight of the ligand molecule.
* SCORE_RB_PEN_NORM_CRT_HEVATOMS: Score for normalized rotamer and heavy heteroatoms with correction.
* SCORE_NORM_CONTACT: Score for normalized contact.
* EVAL: Evaluation value that measures the quality of docking between the ligand and target protein.
* TIME: Time required to perform the docking.

**OPTIONAL-View the file namely ranking.csv file. It is same as bestranking.csv, but for all ligand poses**

In [ ]:
from IPython.display import HTML
import pandas as pd

df = pd.read_csv('/content/results/ranking.csv')

html_table = df.to_html(index=False)
display(HTML(html_table))

**Perhitungan RMSD**

In [ ]:
!obrms -f lig.mol2 /content/results/3HTB_ligand_2_entry_00001_conf_01.mol2 | sed 's/3HTB_ligand_2:3HTB_ligand_2_entry_00001_conf_01/RE-DOCKING/g'

RMSD RE-DOCKING 0.447931



**Visualization**

In [ ]:
!pip install py3Dmol

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!obabel /content/rec.mol2 -O recvis.pdb
!obabel /content/results/3HTB_ligand_2_entry_00001_conf_01.mol2 -O ligvis.pdb

1 molecule converted
1 molecule converted


In [ ]:
import py3Dmol
v = py3Dmol.view()
v.addModel(open('/content/recvis.pdb').read())
v.setStyle({'cartoon':{},'stick':{'radius':.1}})
v.addModel(open('/content/ligvis.pdb').read())
v.setStyle({'model':1},{'stick':{'colorscheme':'dimgrayCarbon','radius':.125}})
v.addModelsAsFrames(open('/content/ligvis.pdb').read())
v.setStyle({'model':1},{'stick':{'colorscheme':'greenCarbon'}})
v.zoomTo({'model':1})
v.rotate(90)

You appear to be running in JupyterLab (or JavaScript failed to load for some other reason). You need to install the 3dmol extension: 
 jupyter labextension install jupyterlab_3dmol

**Optimasi Geometri**




1.  Download Struktur Eugenol di Website Pubchem: https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/CID/3314/record/SDF?record_type=3d&response_type=save&response_basename=Conformer3D_COMPOUND_CID_3314
2. Drag file tersebut ke kiri google colab sehingga, filenya bisa diakses
3. Optimasi geometri dilakukan hanya untuk senyawa yang non native ligand (bahan alam, atau sintesis)






In [27]:
!mv Conformer3D_COMPOUND_CID_3314.sdf comp.sdf

In [28]:
!obminimize -c 1e-6 -cg -h -ff GAFF -n 2500 -cut -rvdw 6.0 -rele 10.0 -pf 10 comp.sdf > comp_opt.mol2


A T O M   T Y P E S

IDX	TYPE	RING
1	os	NO
2	oh	NO
3	ca	AR
4	c3	NO
5	ca	AR
6	ca	AR
7	ca	AR
8	ca	AR
9	ca	AR
10	ce	NO
11	c3	NO
12	c2	NO
13	hc	NO
14	hc	NO
15	ha	NO
16	ha	NO
17	ha	NO
18	ha	NO
19	h1	NO
20	h1	NO
21	h1	NO
22	ho	NO
23	ha	NO
24	ha	NO

C H A R G E S

IDX	CHARGE
1	-0.491450
2	-0.503286
3	-0.039964
4	-0.009611
5	-0.012836
6	0.161530
7	-0.054598
8	0.158451
9	-0.016255
10	-0.086800
11	0.078962
12	-0.102320
13	0.035544
14	0.035544
15	0.065805
16	0.062154
17	0.065524
18	0.056955
19	0.065995
20	0.065995
21	0.065995
22	0.292298
23	0.053184
24	0.053184

S E T T I N G   U P   C A L C U L A T I O N S

SETTING UP BOND CALCULATIONS...
SETTING UP ANGLE CALCULATIONS...
COULD NOT FIND PARAMETERS FOR ANGLE c3-ce-ha, USING DEFAULT PARAMETERS
SETTING UP TORSION CALCULATIONS...
COULD NOT FIND PARAMETERS FOR TORSION ca-c3-ce-c2, USING DEFAULT PARAMETERS
COULD NOT FIND PARAMETERS FOR TORSION ca-c3-ce-ha, USING DEFAULT PARAMETERS
COULD NOT FIND PARAMETERS FOR TORSION hc-c3-ce-c2, USING DEFAULT PARAME

**Dockingkan Senyawa Uji**

In [30]:
!./plants --mode screen /content/dock_comp.txt



                  PLANTS
        Protein-Ligand ANT System
               version 1.2


author: Oliver Korb

scientific contributors: T.E. Exner, T. Stuetzle

contact: Oliver.Korb@uni-konstanz.de


run PLANTS: PLANTS --mode screen yourconfigfile

PLANTS info: CHO hydrogen found 513 HD2 HIS31
PLANTS info: CHO hydrogen found 514 HE1 HIS31
virtual screening progress: 1 of 1
current ligand: /content/comp_opt.mol2 (entry 1)
PLANTS error: Molecule '/content/comp_opt.mol2' contains no atoms!
PLANTS warning: Ligand '/content/comp_opt.mol2' entry 1 skipped.
total virtual screening time: 0.15s
Ligands skipped due to errors: 1


Download file di results dan results 2
